Parsing error after upgrading docker version on Windows

Hi,

I am working on java testcontainers (version 1.14.3) on Windows. recently, my docker desktop version was upgraded from 2.4.0.0 to 2.5.0.0. since then I am getting those exceptions on initialization:

```
Caused by: org.testcontainers.shaded.com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `com.github.dockerjava.api.model.Binds`, **problem: Error parsing Bind 'C:/dev/IM2/BackendTests/Tests/../run/community-eval/logs/:/service/community-eval/logs:rw'**
 at [Source: (org.testcontainers.shaded.okio.RealBufferedSource$1); line: 1, column: 1272] (through reference chain: com.github.dockerjava.api.command.InspectContainerResponse["HostConfig"]->com.github.dockerjava.api.model.HostConfig["Binds"])
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.exc.ValueInstantiationException.from(ValueInstantiationException.java:47) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1735) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:491) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:514) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator._createUsingDelegate(StdValueInstantiator.java:551) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingDelegate(StdValueInstantiator.java:298) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingArrayDelegate(StdValueInstantiator.java:307) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromArray(BeanDeserializerBase.java:1435) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:185) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3259) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.get(OkHttpInvocationBuilder.java:90) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:31) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:13) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.InspectContainerCmdImpl.exec(InspectContainerCmdImpl.java:52) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:428) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325) ~[testcontainers-1.14.3.jar:?]
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81) ~[duct-tape-1.0.8.jar:?]
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323) ~[testcontainers-1.14.3.jar:?]
	... 24 more
**Caused by: java.lang.IllegalArgumentException: Error parsing Bind 'C:/dev/IM2/BackendTests/Tests/../run/community-eval/logs/:/service/community-eval/logs:rw'
	at com.github.dockerjava.api.model.Bind.parse(Bind.java:134) ~[testcontainers-1.14.3.jar:?]**
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_271]
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.ReferencePipeline.toArray(Unknown Source) ~[?:1.8.0_271]
	at com.github.dockerjava.api.model.Binds.fromPrimitive(Binds.java:30) ~[testcontainers-1.14.3.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_271]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_271]
```


It seems that Bind.parse method does not support Windows path at all. 
However, in the previous docker desktop version(2.4.0.0) the Windows path was translated to /host_mnt/C/... which can be parsed by that function. 

any idea how can this be resolved in new docker desktop version?

--- a/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java
+++ b/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java
@@ -33,6 +33,7 @@ import org.testcontainers.utility.DockerImageName;
 import org.testcontainers.utility.DockerLoggerFactory;
 import org.testcontainers.utility.LogUtils;
 import org.testcontainers.utility.MountableFile;
+import org.testcontainers.utility.PathUtils;
 import org.testcontainers.utility.ResourceReaper;
 import org.zeroturnaround.exec.InvalidExitValueException;
 import org.zeroturnaround.exec.ProcessExecutor;
@@ -618,12 +619,13 @@ class ContainerisedDockerCompose extends GenericContainer<ContainerisedDockerCom
         // Map the docker compose file into the container
         final File dockerComposeBaseFile = composeFiles.get(0);
         final String pwd = dockerComposeBaseFile.getAbsoluteFile().getParentFile().getAbsolutePath();
-        final String containerPwd = MountableFile.forHostPath(pwd).getUnixFilesystemPath();
+        final String containerPwd =  convertToUnixFilesystemPath(pwd);
 
         final List<String> absoluteDockerComposeFiles = composeFiles.stream()
             .map(File::getAbsolutePath)
             .map(MountableFile::forHostPath)
-            .map(MountableFile::getUnixFilesystemPath)
+            .map(MountableFile::getFilesystemPath)
+            .map(this::convertToUnixFilesystemPath)
             .collect(toList());
         final String composeFileEnvVariableValue = Joiner.on(UNIX_PATH_SEPERATOR).join(absoluteDockerComposeFiles); // we always need the UNIX path separator
         logger().debug("Set env COMPOSE_FILE={}", composeFileEnvVariableValue);
@@ -668,6 +670,12 @@ class ContainerisedDockerCompose extends GenericContainer<ContainerisedDockerCom
                 " whilst running command: " +
                 StringUtils.join(this.getCommandParts(), ' '));
         }
+    }
+
+    private String convertToUnixFilesystemPath(String path) {
+        return SystemUtils.IS_OS_WINDOWS
+            ? PathUtils.createMinGWPath(path).substring(1)
+            : path;
     }
 }
 

--- a/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java
+++ b/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java
@@ -618,12 +618,12 @@ class ContainerisedDockerCompose extends GenericContainer<ContainerisedDockerCom
         // Map the docker compose file into the container
         final File dockerComposeBaseFile = composeFiles.get(0);
         final String pwd = dockerComposeBaseFile.getAbsoluteFile().getParentFile().getAbsolutePath();
-        final String containerPwd = MountableFile.forHostPath(pwd).getFilesystemPath();
+        final String containerPwd = MountableFile.forHostPath(pwd).getUnixFilesystemPath();
 
         final List<String> absoluteDockerComposeFiles = composeFiles.stream()
             .map(File::getAbsolutePath)
             .map(MountableFile::forHostPath)
-            .map(MountableFile::getFilesystemPath)
+            .map(MountableFile::getUnixFilesystemPath)
             .collect(toList());
         final String composeFileEnvVariableValue = Joiner.on(UNIX_PATH_SEPERATOR).join(absoluteDockerComposeFiles); // we always need the UNIX path separator
         logger().debug("Set env COMPOSE_FILE={}", composeFileEnvVariableValue);

--- a/core/src/main/java/org/testcontainers/containers/GenericContainer.java
+++ b/core/src/main/java/org/testcontainers/containers/GenericContainer.java
@@ -63,6 +63,7 @@ import lombok.Setter;
 import lombok.SneakyThrows;
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.SystemUtils;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.junit.runner.Description;
@@ -944,9 +945,14 @@ public class GenericContainer<SELF extends GenericContainer<SELF>>
      */
     @Override
     public void addFileSystemBind(final String hostPath, final String containerPath, final BindMode mode, final SelinuxContext selinuxContext) {
+        if (SystemUtils.IS_OS_WINDOWS && hostPath.startsWith("/")) {
+            // e.g. Docker socket mount
+            binds.add(new Bind(hostPath, new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
 
-        final MountableFile mountableFile = MountableFile.forHostPath(hostPath);
-        binds.add(new Bind(mountableFile.getResolvedPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
+        } else {
+            final MountableFile mountableFile = MountableFile.forHostPath(hostPath);
+            binds.add(new Bind(mountableFile.getResolvedPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
+        }
     }
 
     /**

--- a/core/src/main/java/org/testcontainers/containers/GenericContainer.java
+++ b/core/src/main/java/org/testcontainers/containers/GenericContainer.java
@@ -946,7 +946,7 @@ public class GenericContainer<SELF extends GenericContainer<SELF>>
     public void addFileSystemBind(final String hostPath, final String containerPath, final BindMode mode, final SelinuxContext selinuxContext) {
 
         final MountableFile mountableFile = MountableFile.forHostPath(hostPath);
-        binds.add(new Bind(mountableFile.getUnixFilesystemPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
+        binds.add(new Bind(mountableFile.getResolvedPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
     }
 
     /**

--- a/core/src/main/java/org/testcontainers/containers/GenericContainer.java
+++ b/core/src/main/java/org/testcontainers/containers/GenericContainer.java
@@ -946,7 +946,7 @@ public class GenericContainer<SELF extends GenericContainer<SELF>>
     public void addFileSystemBind(final String hostPath, final String containerPath, final BindMode mode, final SelinuxContext selinuxContext) {
 
         final MountableFile mountableFile = MountableFile.forHostPath(hostPath);
-        binds.add(new Bind(mountableFile.getResolvedPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
+        binds.add(new Bind(mountableFile.getUnixFilesystemPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
     }
 
     /**

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -105,7 +105,7 @@ public class MountableFile implements Transferable {
      * @return a {@link MountableFile} that may be used to obtain a mountable path
      */
     public static MountableFile forHostPath(@NotNull final String path, Integer mode) {
-        return new MountableFile(new File(path).toURI().toString(), mode);
+        return forHostPath(Paths.get(path), mode);
     }
 
     /**

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -395,6 +395,12 @@ public class MountableFile implements Transferable {
         }
     }
 
+    public String getUnixFilesystemPath() {
+        return SystemUtils.IS_OS_WINDOWS
+            ? PathUtils.createMinGWPath(path).substring(1)
+            : path;
+    }
+
     private int getModeValue(final Path path) {
         int result = Files.isDirectory(path) ? BASE_DIR_MODE : BASE_FILE_MODE;
         return result | this.forcedFileMode;

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -175,7 +175,7 @@ public class MountableFile implements Transferable {
         String result = getResourcePath();
 
         // Special case for Windows
-        if (SystemUtils.IS_OS_WINDOWS && result.startsWith("/")) { // Can this ever happen? It already is a Path.
+        if (SystemUtils.IS_OS_WINDOWS && result.startsWith("/")) {
             // Remove leading /
             result = result.substring(1);
         }

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -105,12 +105,7 @@ public class MountableFile implements Transferable {
      * @return a {@link MountableFile} that may be used to obtain a mountable path
      */
     public static MountableFile forHostPath(@NotNull final String path, Integer mode) {
-        if (SystemUtils.IS_OS_WINDOWS && path.startsWith("/")) {
-            // e.g. Docker socket mount
-            return new MountableFile(path, mode);
-        } else {
-            return forHostPath(Paths.get(path), mode);
-        }
+        return forHostPath(Paths.get(path), mode);
     }
 
     /**

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -395,12 +395,6 @@ public class MountableFile implements Transferable {
         }
     }
 
-    public String getUnixFilesystemPath() {
-        return SystemUtils.IS_OS_WINDOWS
-            ? PathUtils.createMinGWPath(path).substring(1)
-            : path;
-    }
-
     private int getModeValue(final Path path) {
         int result = Files.isDirectory(path) ? BASE_DIR_MODE : BASE_FILE_MODE;
         return result | this.forcedFileMode;

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -105,7 +105,12 @@ public class MountableFile implements Transferable {
      * @return a {@link MountableFile} that may be used to obtain a mountable path
      */
     public static MountableFile forHostPath(@NotNull final String path, Integer mode) {
-        return forHostPath(Paths.get(path), mode);
+        if (SystemUtils.IS_OS_WINDOWS && path.startsWith("/")) {
+            // e.g. Docker socket mount
+            return new MountableFile(path, mode);
+        } else {
+            return forHostPath(Paths.get(path), mode);
+        }
     }
 
     /**
@@ -173,9 +178,6 @@ public class MountableFile implements Transferable {
         if (SystemUtils.IS_OS_WINDOWS && result.startsWith("/")) { // Can this ever happen? It already is a Path.
             // Remove leading /
             result = result.substring(1);
-        } else if (SystemUtils.IS_OS_WINDOWS && result.startsWith("\\")) {
-            // for cases such as Docker socket mounting
-            result = PathUtils.createMinGWPath(result).substring(1);
         }
 
         return result;

--- a/core/src/main/java/org/testcontainers/utility/MountableFile.java
+++ b/core/src/main/java/org/testcontainers/utility/MountableFile.java
@@ -170,9 +170,12 @@ public class MountableFile implements Transferable {
         String result = getResourcePath();
 
         // Special case for Windows
-        if (SystemUtils.IS_OS_WINDOWS && result.startsWith("/")) {
+        if (SystemUtils.IS_OS_WINDOWS && result.startsWith("/")) { // Can this ever happen? It already is a Path.
             // Remove leading /
             result = result.substring(1);
+        } else if (SystemUtils.IS_OS_WINDOWS && result.startsWith("\\")) {
+            // for cases such as Docker socket mounting
+            result = PathUtils.createMinGWPath(result).substring(1);
         }
 
         return result;

